File: radar/web/src/utils/interpolation.js
export function idwInterpolate(x, y, points, power = 2) {
  if (!points.length) return 0;
  let numerator = 0;
  let denominator = 0;
  for (const point of points) {
    const distance = Math.hypot(point.x - x, point.y - y);
    if (distance < 1e-6) return point.value;
    const weight = (point.confidence || 1) / distance ** power;
    numerator += weight * point.value;
    denominator += weight;
  }
  return denominator ? numerator / denominator : 0;
}

File: radar/firmware/src/spatial/interpolation.cpp
#include "spatial/interpolation.h"

#include <cmath>

float IDWInterpolator::interpolate(float x, float y, const std::vector<InterpolationPoint> &points, float power) {
  if (points.empty()) return 0.0f;
  float weighted = 0.0f;
  float sumWeights = 0.0f;
  for (const auto &point : points) {
    const float dx = x - point.x;
    const float dy = y - point.y;
    const float distance = std::sqrt(dx * dx + dy * dy);
    if (distance < 0.0001f) return point.value;
    const float weight = point.confidence / std::pow(distance, power);
    weighted += weight * point.value;
    sumWeights += weight;
  }
  return sumWeights > 0.0f ? weighted / sumWeights : 0.0f;
}

float IDWInterpolator::confidence(float x, float y, const std::vector<InterpolationPoint> &points) {
  if (points.empty()) return 0.0f;
  float nearest = 9999.0f;
  for (const auto &point : points) {
    const float dx = x - point.x;
    const float dy = y - point.y;
    nearest = std::min(nearest, std::sqrt(dx * dx + dy * dy));
  }
  return 1.0f / (1.0f + nearest);
}
